packages/react-cache/src/__tests__/ReactCache-test.internal.js

```diff
@@ -0,0 +1,327 @@
+    root.update(<App />);
+    expect(root).toFlushAndThrow('Failed to load: Hi');
+    expect(ReactTestRenderer).toHaveYielded(['Error! [Hi]', 'Error! [Hi]']);
+  });
+
+  it('warns if non-primitive key is passed to a resource without a hash function', () => {
+    const BadTextResource = createResource(([text, ms = 0]) => {
+      return new Promise((resolve, reject) =>
+        setTimeout(() => {
+          resolve(text);
+        }, ms),
+      );
+    });
+
+    function App() {
+      ReactTestRenderer.unstable_yield('App');
+      return BadTextResource.read(['Hi', 100]);
+    }
+
+    const root = ReactTestRenderer.create(
+      <Suspense>
+        <App />
+      </Suspense>,
+      {
+        unstable_isConcurrent: true,
+      },
+    );
+
+    if (__DEV__) {
+      expect(() => {
+        expect(root).toFlushAndYield(['App']);
+      }).toWarnDev(
+        [
+          'Invalid key type. Expected a string, number, symbol, or ' +
+            'boolean, but instead received: Hi,100\n\n' +
+            'To use non-primitive values as keys, you must pass a hash ' +
+            'function as the second argument to createResource().',
+        ],
+        {withoutStack: true},
+      );
+    } else {
+      expect(root).toFlushAndYield(['App']);
+    }
+  });
+
+  it('evicts least recently used values', async () => {
+    ReactCache.unstable_setGlobalCacheLimit(3);
+
+    // Render 1, 2, and 3
+    const root = ReactTestRenderer.create(
+      <Suspense>
+        <AsyncText ms={100} text={1} />
+        <AsyncText ms={100} text={2} />
+        <AsyncText ms={100} text={3} />
+      </Suspense>,
+      {
+        unstable_isConcurrent: true,
+      },
+    );
+    expect(root).toFlushAndYield([
+      'Suspend! [1]',
+      'Suspend! [2]',
+      'Suspend! [3]',
+    ]);
+    jest.advanceTimersByTime(100);
+    expect(ReactTestRenderer).toHaveYielded([
+      'Promise resolved [1]',
+      'Promise resolved [2]',
+      'Promise resolved [3]',
+    ]);
+    expect(root).toFlushAndYield([1, 2, 3]);
+    expect(root).toMatchRenderedOutput('123');
+
+    // Render 1, 4, 5
+    root.update(
+      <Suspense>
+        <AsyncText ms={100} text={1} />
+        <AsyncText ms={100} text={4} />
+        <AsyncText ms={100} text={5} />
+      </Suspense>,
+    );
+
+    expect(root).toFlushAndYield([1, 'Suspend! [4]', 'Suspend! [5]']);
+    jest.advanceTimersByTime(100);
+    expect(ReactTestRenderer).toHaveYielded([
+      'Promise resolved [4]',
+      'Promise resolved [5]',
+    ]);
+    expect(root).toFlushAndYield([1, 4, 5]);
+    expect(root).toMatchRenderedOutput('145');
+
+    // We've now rendered values 1, 2, 3, 4, 5, over our limit of 3. The least
+    // recently used values are 2 and 3. They will be evicted during the
+    // next sweep.
+    evictLRU();
+
+    root.update(
+      <Suspense>
+        <AsyncText ms={100} text={1} />
+        <AsyncText ms={100} text={2} />
+        <AsyncText ms={100} text={3} />
```

Comment:
Should we use or add a PureComponent of AsyncText case like :

```
<AsyncText ms={100} text={2} />
<AsyncText ms={100} text={3} />
<AsyncText ms={100} text={1} />
```
here? IMO, we need a case which just test the cache read.

Reply:
This test does read from the cache? And idk what you mean by "common reconciliation"

Reply:
Sorry. I mean, this is just test the function component(the `AsyncText`) but in real world, we have class component which can be `PureComponent` or normal `Component`, so in the common reconciliation, they will go to different switch case or codes, I think it maybe should be tested too.